src/libinjector/win_injector.c

```diff
@@ -1065,7 +1140,196 @@ static event_response_t injector_int3_cb(drakvuf_t drakvuf, drakvuf_trap_info_t*
+            }
+        }
+
+        PRINT_DEBUG("Writing file...\n");
+        amount = fread(buf + FILE_BUF_RESERVED, 1, FILE_BUF_SIZE - FILE_BUF_RESERVED, injector->host_file);
+        PRINT_DEBUG("Amount: %lx\n", amount);
+
+        if (!amount)
+        {
+            PRINT_DEBUG("Finishing\n");
+
+            if (!setup_close_handle_stack(injector, info))
+            {
+                PRINT_DEBUG("Failed to setup stack for closing handle\n");
+                return 0;
+            }
+
+            info->regs->rip = injector->close_handle;
+
+            injector->status = STATUS_CLOSE_FILE_OK;
+
+            return VMI_EVENT_RESPONSE_SET_REGISTERS;
+        }
+        else
+        {
+            PRINT_DEBUG("Writing...\n");
+
+            if (!setup_write_file_stack(injector, info, amount))
+            {
+                PRINT_DEBUG("Failed to setup stack for passing inputs!\n");
+                return 0;
+            }
+        }
+
+        access_context_t ctx = { 0 };
+        ctx.translate_mechanism = VMI_TM_PROCESS_DTB;
+        ctx.dtb = info->regs->cr3;
+        ctx.addr = injector->payload_addr + FILE_BUF_RESERVED;
+
+        vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+        vmi_write(vmi, &ctx, amount, buf + FILE_BUF_RESERVED, NULL);
+        drakvuf_release_vmi(drakvuf);
+
+        info->regs->rip = injector->write_file;
+
+        injector->status = STATUS_WRITE_FILE_OK;
+
+        return VMI_EVENT_RESPONSE_SET_REGISTERS;
+    }
+
+    if (!injector->is32bit &&
+        STATUS_CREATE_FILE_OK == injector->status &&
+        INJECT_METHOD_READ_FILE == injector->method)
+    {
+        PRINT_DEBUG("File create result %lx\n", info->regs->rax);
+        injector->file_handle = info->regs->rax;
+        injector->host_file = fopen(injector->binary_path, "wb");
+
+        PRINT_DEBUG("Reading file...\n");
+
+        if (!setup_read_file_stack(injector, info))
+        {
+            PRINT_DEBUG("Failed to setup stack for passing inputs!\n");
+            return 0;
+        }
+
+        info->regs->rip = injector->read_file;
+
+        injector->status = STATUS_READ_FILE_OK;
+
+        return VMI_EVENT_RESPONSE_SET_REGISTERS;
+    }
+
+    if (!injector->is32bit && STATUS_READ_FILE_OK == injector->status)
+    {
+        uint8_t buf[FILE_BUF_SIZE];
+
+        PRINT_DEBUG("File read result: %lx\n", info->regs->rax);
+
+        access_context_t ctx = { 0 };
+        ctx.translate_mechanism = VMI_TM_PROCESS_DTB;
+        ctx.dtb = info->regs->cr3;
+        ctx.addr = injector->payload_addr;
+
+        vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+        vmi_read(vmi, &ctx, FILE_BUF_SIZE, buf, NULL);
+        drakvuf_release_vmi(drakvuf);
+
+        uint32_t* num_bytes = (uint32_t*)buf;
+
+        if (*num_bytes > FILE_BUF_SIZE)
+        {
+            *num_bytes = FILE_BUF_SIZE;
+            PRINT_DEBUG("Number of bytes read by ReadFile is greater than the buffer size, truncating.\n");
```

Comment:
Is this a safe "default" behavior?

Reply:
I think so. In any case we will not read more than the allocated buffer.

Reply:
Ok, the proper question is: why do we have a static buffer with random amount, instead dynamically allocated one?

Reply:
This is a read/write buffer which is dynamically allocated inside guest and then it's mirrored with an array of bytes on host. We do need some arbitrary size here, otherwise we would end up allocating 4 GB buffers to read 4 GB files etc.

Reply:
In that case we should keep reading the file in batches to get entire file instead of truncating.

Reply:
It is done so already. This check is to ensure that we are not trolled by ReadFile saying that it read e.g. >1 MB into 1 MB buffer. It might happen in practice in case ReadFile gets somehow altered.